internal/backend/sessions/sessioncalls/calls.go

```diff
@@ -85,34 +132,57 @@ func (cs *calls) Call(ctx workflow.Context, params *CallParams) (sdktypes.Sessio
 
 	goCtx := temporalclient.NewWorkflowContextAsGOContext(ctx)
 
-	if params.ForceInternal || fnv.GetFunction().HasFlag(sdktypes.PureFunctionFlag) {
+	if params.ForceInternal || fnvf.HasFlag(sdktypes.PureFunctionFlag) {
 		z.Debug("function call outside activity")
 
 		var err error
 
-		if fnv.GetFunction().HasFlag(sdktypes.PrivilidgedFunctionFlag) {
+		if fnvf.HasFlag(sdktypes.PrivilidgedFunctionFlag) {
 			goCtx = sessioncontext.WithWorkflowContext(goCtx, ctx)
 		}
 
 		if _, attempt, err = cs.executeCall(goCtx, params.SessionID, seq, params.Poller, params.Executors); err != nil {
 			return sdktypes.NewSessionCallAttemptResult(sdktypes.InvalidValue, fmt.Errorf("internal call: %w", err)), nil
 		}
 	} else {
-		executorsForSessions[params.SessionID.String()] = params.Executors // HACK
+		xid := fnvf.ExecutorID()
 
-		actx := workflow.WithActivityOptions(
-			ctx,
-			workflow.ActivityOptions{
+		// neither a non-integration (like a run) or an integration that is builtin.
+		local := !xid.IsIntegrationID() || fixtures.IsBuiltinIntegrationID(xid.ToIntegrationID())
+
+		if local {
+			executorsForSessions[params.SessionID.String()] = params.Executors
+			defer func() { delete(executorsForSessions, params.SessionID.String()) }()
+		}
+
+		for retry := true; retry; {
+			aopts := workflow.ActivityOptions{
 				TaskQueue:              taskQueueName,
 				ActivityID:             fmt.Sprintf("session_call_%s_%d", params.SessionID.Value(), seq),
 				ScheduleToCloseTimeout: cs.config.Temporal.ActivityScheduleToCloseTimeout,
 				HeartbeatTimeout:       cs.config.Temporal.ActivityHeartbeatTimeout,
-			},
-		)
+			}
 
-		ret := callActivityOutputs{Retry: true}
+			if local {
+				// When a local execution is required, it is scheduled on a unique worker. These local executions
+				// need access to the state that is built by the workflow.
+
+				if aopts.TaskQueue = cs.uniqueWorkerCallTaskQueueName; aopts.TaskQueue == "" {
+					return sdktypes.InvalidSessionCallAttemptResult, errors.New("local session worker for activities is not registerd")
+				}
+
+				aopts.ScheduleToStartTimeout = cs.config.Temporal.ActivityScheduleToStartTimeout
+				if aopts.ScheduleToStartTimeout == 0 {
+					z.Warn("call activity schedule to start timeout is 0 and local exec is needed")
```

Comment:
Tiny readability nit:
```suggestion
					z.Warn("call activity schedule-to-start timeout is 0 and local exec is needed")
```